In [1]:
using JuMP;
using GLPK;

# Questão 13

### Restrições do problema

In [2]:
coeff = [6, 4, 4, 1, 1];
n = size(coeff, 1);

A = [2 2 3 1 2];

b = [7];
s = ["="];

### Solução Inicial

In [3]:
model = Model(GLPK.Optimizer);
@variable(model, x[i = 1:n] >= 0, base_name = "x");
opt_function = @expression(model, coeff'*x);
@constraint(model, C, A*x .<= b);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar = value.(x);
println(xstar, ' ', JuMP.objective_value(model));

[3.5, 0.0, 0.0, 0.0, 0.0] 21.0


### Restrição $x_1 \leq 3$

In [4]:
model = Model(GLPK.Optimizer);
@variable(model, x[i = 1:n] >= 0, base_name = "x");
opt_function = @expression(model, coeff'*x);
@constraint(model, C1, A*x .<= b);
@constraint(model, C2, x[1] <= 3);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar = value.(x);
println(xstar, ' ', JuMP.objective_value(model));

[3.0, 0.5, 0.0, 0.0, 0.0] 20.0


### Restrição $x_1 \geq 4$

In [5]:
model = Model(GLPK.Optimizer);
@variable(model, x[i = 1:n] >= 0, base_name = "x");
opt_function = @expression(model, coeff'*x);
@constraint(model, C1, A*x .<= b);
@constraint(model, C2, x[1] >= 4);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar = value.(x);
println(xstar, ' ', JuMP.objective_value(model));

[3.5, 0.0, 0.0, 0.0, 0.0] 21.0


Ou seja, é infactível, podendo parar de expandir esse ramo.

### Restrições $x_1 \leq 3$, $x_2 \leq 0$

In [6]:
model = Model(GLPK.Optimizer);
@variable(model, x[i = 1:n] >= 0, base_name = "x");
opt_function = @expression(model, coeff'*x);
@constraint(model, C1, A*x .<= b);
@constraint(model, C2, x[1] <= 3);
@constraint(model, C3, x[2] <= 0);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar = value.(x);
println(xstar, ' ', JuMP.objective_value(model));

[3.0, 0.0, 0.3333333333333333, 0.0, 0.0] 19.333333333333332


Onde já temos as restrições de $x_1$ e $x_2$ inteiros satisfeitas, ou seja, podemos parar de expandir esse ramo.

### Restrições $x_1 \leq 3$, $x_2 \geq 1$

In [7]:
model = Model(GLPK.Optimizer);
@variable(model, x[i = 1:n] >= 0, base_name = "x");
opt_function = @expression(model, coeff'*x);
@constraint(model, C1, A*x .<= b);
@constraint(model, C2, x[1] <= 3);
@constraint(model, C3, x[2] >= 1);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar = value.(x);
println(xstar, ' ', JuMP.objective_value(model));

[2.5, 1.0, 0.0, 0.0, 0.0] 19.0


Note que aqui não temos as restrições de inteiro satisfeitas, mas o valor de $z$ é inferior a $19.3333$, que foi obtido no outro ramo, ou seja, podemos parar de expandir esse ramo, chegando a solução $x_1 = 3$, $x_2 = 0$, $x_3 = \frac{1}{3}$, $x_4 = 0$, $x_5 = 0$ e $z = \frac{58}{3}$.

<img src = "q13bab.png">

## Gabarito

In [8]:
coeff = [6, 4, 4, 1, 1];
n = size(coeff, 1);

A = [2 2 3 1 2];

b = [7];
s = ["="];

model = Model(GLPK.Optimizer);
@variable(model, x1[i = 1:2] >= 0, base_name = "x", Int);
@variable(model, x2[i = 1:3] >= 0, base_name = "x");
opt_function = @expression(model, coeff[1:2]'*x1 + coeff[3:n]'*x2);
@constraint(model, C, A[1, 1:2]'*x1 + A[1, 3:n]'*x2 .<= b);
@objective(model, Max, opt_function);

status = JuMP.optimize!(model);
xstar1 = value.(x1);
xstar2 = value.(x2);
println(xstar1)
println(xstar2)
println(JuMP.objective_value(model));

[3.0, 0.0]
[0.3333333333333333, 0.0, 0.0]
19.333333333333332
